broker/client/client.go

```diff
@@ -712,3 +514,276 @@ func getPeerInfo(peer *ill_db.Peer, symbol string) (string, iso18626.TypeAgencyI
+	}
+	transCtx.selectedSupplier = selected
+	transCtx.selectedPeer = supplier
+	return transCtx, nil
+}
+
+func (c *Iso18626Client) createAndSendSupplyingAgencyMessage(ctx extctx.ExtendedContext, event events.Event) (events.EventStatus, *events.EventResult) {
+	trCtx, err := c.readTransactionContext(ctx, event, false)
+	if err != nil {
+		return events.LogErrorAndReturnResult(ctx, FailedToGetTrCtx, err)
+	}
+
+	msgTarget, err := c.determineMessageTarget(ctx, trCtx)
+	if err != nil {
+		if msgTarget != nil && msgTarget.problemDetails != nil {
+			return events.LogProblemAndReturnResult(ctx, InvalidState, msgTarget.problemDetails.message, nil)
+		}
+		return events.LogErrorAndReturnResult(ctx, FailedToDetermineTarget, err)
+	}
+
+	if msgTarget.status == "" {
+		resData := events.EventResult{}
+		resData.Note = msgTarget.note
+		return events.EventStatusSuccess, &resData
+	}
+
+	message := buildSupplyingAgencyMessage(trCtx, msgTarget)
+
+	return c.sendAndUpdateStatus(ctx, trCtx, message)
+}
+
+func (c *Iso18626Client) determineMessageTarget(ctx extctx.ExtendedContext, trCtx transactionContext) (*messageTarget, error) {
+	msgSupplierSymbol := getSymbol(trCtx.event.EventData.IncomingMessage)
+	isNotification := isNotification(trCtx.event.EventData.IncomingMessage)
+	if trCtx.selectedSupplier == nil {
+		return c.handleNoSelectedSupplier(ctx, trCtx, msgSupplierSymbol, isNotification)
+	}
+
+	if isNotification && msgSupplierSymbol != "" && msgSupplierSymbol != trCtx.selectedSupplier.SupplierSymbol {
+		return c.handleSkippedSupplierNotification(ctx, trCtx, msgSupplierSymbol)
+	}
+
+	return handleSelectedSupplier(trCtx)
+}
+
+func (c *Iso18626Client) handleNoSelectedSupplier(ctx extctx.ExtendedContext, trCtx transactionContext, msgSupplierSymbol string, isNotification bool) (*messageTarget, error) {
+	if isNotification && msgSupplierSymbol != "" {
+		return c.handleSkippedSupplierNotification(ctx, trCtx, msgSupplierSymbol)
+	}
+
+	return &messageTarget{status: iso18626.TypeStatusUnfilled, firstMessage: false}, nil
+}
+
+func (c *Iso18626Client) handleSkippedSupplierNotification(ctx extctx.ExtendedContext, trCtx transactionContext, msgSupplierSymbol string) (*messageTarget, error) {
+	skipped, skippedPeer, err := c.getSkippedSupplierAndPeer(ctx, trCtx.transaction.ID, msgSupplierSymbol)
+	if err != nil {
+		return nil, err
+	}
+
+	if trCtx.requester.BrokerMode == string(extctx.BrokerModeTransparent) {
+		return &messageTarget{
+			supplier:     skipped,
+			peer:         skippedPeer,
+			status:       iso18626.TypeStatus(trCtx.transaction.LastSupplierStatus.String),
+			firstMessage: false,
+		}, nil
+	}
+
+	return &messageTarget{
+		problemDetails: &struct{ message string }{
+			message: fmt.Sprintf("ignored notification from skipped supplier %s due to requester mode %s", msgSupplierSymbol, trCtx.requester.BrokerMode),
+		},
+	}, nil
+}
+
+func handleSelectedSupplier(trCtx transactionContext) (*messageTarget, error) {
```

Comment:
@JanisSaldabols would you mind putting back some comments on the return value so we know in which cases this happens?